client-spark/common/src/main/java/org/apache/spark/shuffle/writer/DataPusher.java

```diff
@@ -0,0 +1,124 @@
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import com.google.common.collect.Queues;
+import com.google.common.collect.Sets;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.uniffle.client.api.ShuffleWriteClient;
+import org.apache.uniffle.client.response.SendShuffleDataResult;
+import org.apache.uniffle.common.ShuffleBlockInfo;
+import org.apache.uniffle.common.util.ThreadUtils;
+
+public class DataPusher implements Closeable {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DataPusher.class);
+
+  private final ExecutorService executorService;
+
+  private final ShuffleWriteClient shuffleWriteClient;
+  private final Map<String, Set<Long>> taskToSuccessBlockIds;
+  private final Map<String, Set<Long>> taskToFailedBlockIds;
+  private String appId;
+  private final Set<String> failedTaskIds;
+
+  public DataPusher(ShuffleWriteClient shuffleWriteClient,
+      Map<String, Set<Long>> taskToSuccessBlockIds,
+      Map<String, Set<Long>> taskToFailedBlockIds,
+      Set<String> failedTaskIds,
+      int threadPoolSize,
+      int threadKeepAliveTime) {
+    this.shuffleWriteClient = shuffleWriteClient;
+    this.taskToSuccessBlockIds = taskToSuccessBlockIds;
+    this.taskToFailedBlockIds = taskToFailedBlockIds;
+    this.failedTaskIds = failedTaskIds;
+    this.executorService = new ThreadPoolExecutor(
+        threadPoolSize,
+        threadPoolSize * 2,
+        threadKeepAliveTime,
+        TimeUnit.SECONDS,
+        Queues.newLinkedBlockingQueue(Integer.MAX_VALUE),
+        ThreadUtils.getThreadFactory(this.getClass().getName())
+    );
+  }
+
+  public CompletableFuture<Long> send(AddBlockEvent event) {
+    assert appId != null;
+    return CompletableFuture.supplyAsync(() -> {
+      String taskId = event.getTaskId();
+      List<ShuffleBlockInfo> shuffleBlockInfoList = event.getShuffleDataInfoList();
+      try {
+        SendShuffleDataResult result = shuffleWriteClient.sendShuffleData(
+            appId,
+            shuffleBlockInfoList,
+            () -> !isValidTask(taskId)
+        );
+        putBlockId(taskToSuccessBlockIds, taskId, result.getSuccessBlockIds());
+        putBlockId(taskToFailedBlockIds, taskId, result.getFailedBlockIds());
+      } finally {
+        List<Runnable> callbackChain = Optional.of(event.getProcessedCallbackChain()).orElse(Collections.EMPTY_LIST);
+        for (Runnable runnable : callbackChain) {
+          runnable.run();
+        }
+      }
+      return shuffleBlockInfoList.stream()
+          .map(x -> x.getFreeMemory())
+          .reduce((a, b) -> a + b)
+          .get();
+    }, executorService);
+  }
+
+  private synchronized void putBlockId(
+      Map<String, Set<Long>> taskToBlockIds,
+      String taskAttemptId,
+      Set<Long> blockIds) {
+    if (blockIds == null || blockIds.isEmpty()) {
+      return;
+    }
+    taskToBlockIds.putIfAbsent(taskAttemptId, Sets.newConcurrentHashSet());
```

Comment:
Could `computeIfAbsent` merge these two lines(106-107) of code?

Reply:
👍. This similar optimization could create an issue to track.